File: location-server/src/main/java/nl/lucien/domain/TimedLocation.java
package nl.lucien.domain;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import lombok.Builder;
import lombok.Data;

/**
 * Represents a point on Earth at a certain time.
 */
@JsonDeserialize(builder = Path.Builder.class)
@Builder
@Data
public class TimedLocation {

    private Location location;
    private Long timestamp; // Unix timestamp (ms)

    @JsonPOJOBuilder(withPrefix = "")
    public static class Builder {
    }
}

File: location-tracker/src/main/kotlin/location/tracker/domain/TimedLocation.kt
package location.tracker.domain

/**
 * Represents a location and a moment in time.
 */
data class TimedLocation(
    val longitude: Double,
    val latitude: Double,
    val timestamp: Long
) {
    init {
        if( longitude < -180 || longitude > 180 ) {
            throw IllegalArgumentException("longitude out of range")
        } else if( latitude < -90 || latitude > 90 ) {
            throw IllegalArgumentException("latitude out of range")
        } else if( timestamp < 0 ) {
            throw IllegalArgumentException("timestamp less than 0")
        }
    }

    override fun toString(): String {
        return "TimedLocation(longitude=$longitude, latitude=$latitude, timestamp=$timestamp)"
    }

}
